staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -768,6 +770,126 @@ func (r *Request) Watch(ctx context.Context) (watch.Interface, error) {
 	}
 }
 
+type WatchListResult struct {
+	// err holds any errors we might have received
+	// during streaming.
+	err error
+
+	// items hold the collected data
+	items []runtime.Object
+
+	// initialEventsEndBookmarkRV holds the resource version
+	// extracted from the bookmark event that marks
+	// the end of the stream.
+	initialEventsEndBookmarkRV string
+
+	// gv represents the API version
+	// it is used to construct the final list response
+	// normally this information is filled by the server
+	gv schema.GroupVersion
+}
+
+func (r WatchListResult) Into(obj runtime.Object) error {
+	if r.err != nil {
+		return r.err
+	}
+
+	listPtr, err := meta.GetItemsPtr(obj)
+	if err != nil {
+		return err
+	}
+	listVal, err := conversion.EnforcePtr(listPtr)
+	if err != nil {
+		return err
+	}
+	if listVal.Kind() != reflect.Slice {
+		return fmt.Errorf("need a pointer to slice, got %v", listVal.Kind())
+	}
+
+	if len(r.items) == 0 {
+		listVal.Set(reflect.MakeSlice(listVal.Type(), 0, 0))
```

Comment:
The apiserver always returns an empty list instead of `nil`. In JSON it is `"items":[]` and not `"items":null`.
Since we are computing the list manually we need to do the same. If we don't do it then on the client-side before accessing the `items` we would have to check if the array has been initialised.

Reply:
That's how I remember it, but I wanted to see the code that does that :)

Reply:
Two places I know about on the server-side are :
- https://github.com/kubernetes/kubernetes/blob/91655922197276ced2e4e36114bb753e384dbd83/staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go#L797
- https://github.com/kubernetes/kubernetes/blob/d8bd150784bb4825ae891dd0ec84625bdba0f2b8/staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go#L822

which give us `"items":[]` which is then unmarshalled into `Items: []type` and not into `Items: nil`

Reply:
OK - thanks for double checking this!